File: MacOS/Type4Me/ASR/Providers/OpenAIASRConfig.swift
import Foundation

struct OpenAIASRConfig: ASRProviderConfig, Sendable {

    static let provider = ASRProvider.openai
    static let displayName = "OpenAI Whisper"

    static let credentialFields: [CredentialField] = [
        CredentialField(key: "apiKey", label: "API Key", placeholder: "sk-...", isSecure: true, isOptional: false, defaultValue: ""),
        CredentialField(key: "baseURL", label: "Base URL", placeholder: "https://api.openai.com/v1", isSecure: false, isOptional: true, defaultValue: "https://api.openai.com/v1"),
    ]

    let apiKey: String
    let baseURL: String

    init?(credentials: [String: String]) {
        guard let key = credentials["apiKey"], !key.isEmpty else { return nil }
        self.apiKey = key
        self.baseURL = credentials["baseURL"]?.isEmpty == false
            ? credentials["baseURL"]!
            : "https://api.openai.com/v1"
    }

    func toCredentials() -> [String: String] {
        ["apiKey": apiKey, "baseURL": baseURL]
    }

    var isValid: Bool { !apiKey.isEmpty }
}

File: Windows/Type4Me/ASR/Providers/OpenAIASRConfig.cs
namespace Type4Me.ASR.Providers;

public sealed class OpenAIASRConfig
{
    public string ApiKey { get; }
    public string BaseURL { get; }

    public static ASRProvider Provider => ASRProvider.OpenAI;
    public static string DisplayName => "OpenAI Whisper";

    public static CredentialField[] CredentialFields =>
    [
        new() { Key = "apiKey", Label = "API Key", Placeholder = "sk-...", IsSecure = true },
        new() { Key = "baseURL", Label = "Base URL", Placeholder = "https://api.openai.com/v1", IsSecure = false, IsOptional = true, DefaultValue = "https://api.openai.com/v1" },
    ];

    private OpenAIASRConfig(string apiKey, string baseURL) { ApiKey = apiKey; BaseURL = baseURL; }

    public static OpenAIASRConfig? TryCreate(Dictionary<string, string> credentials)
    {
        if (!credentials.TryGetValue("apiKey", out var key) || string.IsNullOrEmpty(key)) return null;
        var baseURL = credentials.GetValueOrDefault("baseURL");
        if (string.IsNullOrEmpty(baseURL)) baseURL = "https://api.openai.com/v1";
        return new OpenAIASRConfig(key, baseURL);
    }

    public Dictionary<string, string> ToCredentials() => new() { ["apiKey"] = ApiKey, ["baseURL"] = BaseURL };
    public bool IsValid => !string.IsNullOrEmpty(ApiKey);
}
